### PR TITLE
metrics for CEL compilation and evaluation

### PR DESCRIPTION
**Pull request by jiahuif**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR adds timing metrics for CEL-related operations, namely compilation and evaluation. These metrics are provisional and used only for an indication of CEL performance.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go [hunk 1]
```diff
@@ -187,7 +189,9 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			// transition rules are evaluated only if there is a comparable existing value
 			continue
 		}
+		t := time.Now()
 		evalResult, evalDetails, err := compiled.Program.ContextEval(ctx, activation)
+		metrics.Metrics.ObserveEvaluation(time.Since(t))
```
[the remark this answers is not in the document]
Comment (jiahuif) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:194]: I believe this is the only place where we do the evaluation for now. We can keep an eye of new code where we refer to `ContextEval`

[the remark this answers is not in the document]
Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go]: We agreed to move upper to monitor the entire validation process. Thank you Jiahui for addressing it!

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"time"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// TODO(jiahuif) CEL is to be used in multiple components, revise naming when that happens.
+const (
+	subsystem = "cel"
```
[the remark this answers is not in the document]
Comment (jiahuif) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/metrics/metrics.go:29]: There is little point to make it long. Would you like my idea of adding the namespace, like above?

[the remark this answers is not in the document]
Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/metrics/metrics.go:29]: Do you mean that we would share the namespace in the future cel related metrics like in admission control?

[the remark this answers is not in the document]
Comment (jiahuif) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/metrics/metrics.go:29]: Yes. Let's do that for now. We can change it when we make more progress with CEL and graduating this metric.

**Comment by cici37** [comment]
/assign @deads2k 
Would you mind to take a look when have time? Thank you

**Comment by dashpole** [comment]
/assign @logicalhan 
(for sig-instrumentation)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/metrics/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (logicalhan) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/metrics/metrics.go:29]: I prefer 'cel' here for whatever it's worth, it's not super likely to collide with admission control metrics that are cel related.

**Review by logicalhan — COMMENTED** [review]
/lgtm
/approve

**Comment by deads2k** [comment]
It looks like a good starting point.  I suspect we may end up needing to slice this based on which CR is being evaluated, but the locations and granularity of measurement look good.

/approve

**Comment by cici37** [comment]
@liggitt Would you mind to help with the dep approval on this one? Thank you 😊

**Comment by liggitt** [comment]
/approve
for modules.txt update

**Pull request merged**